services/rsources/stats_collector.go

```diff
@@ -0,0 +1,152 @@
+package rsources
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	uuid "github.com/gofrs/uuid"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/tidwall/gjson"
+)
+
+// StatsCollector collects and publishes stats as jobs are
+// being created, processed and their statuses are being updated.
+type StatsCollector interface {
```

Comment:
Sadly I cannot follow a best practice unless I see its value during application.
For the specific design, the package in question, `rsources`, doesn't export any struct, it only exports an interface, `StatsCollector` and a producer method which produces the `StatsCollector`. So I am neither using an interface for _common behaviour_ nor for _restricting behaviour_, I am merely using an interface for **_decoupling_** the code from actual implementation.

My interface takes the place of a struct, hiding all implementation details. It would be an _interface pollution_ if I was exporting both an interface and a struct.

Finally, if the consumer side wants to declare their own unexported interface for restricting the behaviour of `rsources.StatsCollector` I am fine with it, go ahead, although I wouldn't consider it necessary myself, since it looks like having a coherent set of methods already and I don't see any harm by not restricting them :)

Reply:
> Edit: I would only consider exporting a struct from a package in case its zero-value would make sense to the consumer. In this specific case it doesn't, since stats collector has dependencies of its own that need to be satisfied in order to be usable.

You can either return an unexported `struct` forcing devs to pass via your constructor or use `sync.Once` to init things. Both are viable options, I sometimes return unexported structs in some cases and it worked well for me.

Reply:
> You can either return an unexported `struct`

I would never consider returning an unexported type from an exported method, regardless if this is an interface or struct.
What is exported represents the "public" API of the package, it should be well-defined and well-documented. Since godoc, or any other language's doc tool, doesn't include unexported structs, types or interfaces, it would be very confusing for a consumer to have to guess what is it that he got after calling a method :)